drivers/i2c/i2c_mspm0.c

```diff
@@ -0,0 +1,859 @@
+
+	if(data->state == I2C_MSPM0_TIMEOUT){
+		return -ETIMEDOUT;
+	}
+
+	/* If error, return error */
+	if (((DL_I2C_getControllerStatus(config->base) & DL_I2C_CONTROLLER_STATUS_ERROR) != 0x00) ||
+		(data->state == I2C_MSPM0_ERROR)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int i2c_mspm0_transmit(const struct device *dev, struct i2c_msg msg, uint16_t addr)
+{
+	const struct i2c_mspm0_config *config = dev->config;
+	struct i2c_mspm0_data *data = dev->data;
+
+	/* Sending address without data is not supported */
+	if (msg.len == 0) {
+		return -EIO;
+	}
+
+	/* Update cached msg and addr */
+	data->msg = msg;
+	data->addr = addr;
+
+	/* Update the state */
+	data->state = I2C_MSPM0_IDLE;
+
+	/* Flush anything that is left in the stale FIFO */
+	DL_I2C_flushControllerTXFIFO(config->base);
+
+	/*
+	 * Fill the FIFO
+	 *  The FIFO is 8-bytes deep, and this function will return number
+	 *  of bytes written to FIFO
+	 */
+	data->count =
+		DL_I2C_fillControllerTXFIFO(config->base, data->msg.buf, data->msg.len);
+
+	/* Enable TXFIFO trigger interrupt if there are more bytes to send */
+	if (data->count < data->msg.len) {
+		DL_I2C_enableInterrupt(config->base,
+					DL_I2C_INTERRUPT_CONTROLLER_TXFIFO_TRIGGER);
+	} else {
+		DL_I2C_disableInterrupt(config->base,
+					DL_I2C_INTERRUPT_CONTROLLER_TXFIFO_TRIGGER);
+	}
+
+	/*
+	 * Send the packet to the controller.
+	 * This function will send Start + Stop automatically
+	 */
+	data->state = I2C_MSPM0_TX_STARTED;
+	while (!(DL_I2C_getControllerStatus(config->base) &
+		 DL_I2C_CONTROLLER_STATUS_IDLE))
+		;
+
+	DL_I2C_startControllerTransfer(config->base, data->addr,
+					DL_I2C_CONTROLLER_DIRECTION_TX, data->msg.len);
+
+	while (!(DL_I2C_getControllerStatus(config->base) & DL_I2C_CONTROLLER_STATUS_IDLE));
+
+	if(data->state == I2C_MSPM0_TIMEOUT){
+		return -ETIMEDOUT;
+	}
+
+	/* If error, return error */
+	if (((DL_I2C_getControllerStatus(config->base) & DL_I2C_CONTROLLER_STATUS_ERROR) != 0x00) ||
+		(data->state == I2C_MSPM0_ERROR)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int i2c_mspm0_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+				   uint16_t addr)
+{
+	struct i2c_mspm0_data *data = dev->data;
+	uint32_t config;
+	int ret = 0;
+
+	/* Sending address with no data not supported */
+	if (num_msgs == 0) {
+		return -EINVAL;
+	}
+
+	k_sem_take(&data->i2c_busy_sem, K_FOREVER);
+
+	if (data->is_target) {
+		/* currently target is registered. Controller is disabled */
+		k_sem_give(&data->i2c_busy_sem);
+		return -EBUSY;
+	}
+
+	/* Transmit each message */
+	for (int i = 0; i < num_msgs; i++) {
+		if ((msgs[i].flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) {
+			ret = i2c_mspm0_transmit(dev, msgs[i], addr);
+		} else {
+			ret = i2c_mspm0_receive(dev, msgs[i], addr);
+		}
+	}
+
+	if(ret == -ETIMEDOUT){
+		i2c_mspm0_get_config(dev, &config);
+		i2c_mspm0_reset_peripheral_controller(dev);
+	}
+
```

Comment:
out:

(Idea behing this label to release the semaphore is to minimize the risk to not give it back, so it takes is at one place, and gives it at a unique place as well. At least whenever possible. Easier to read, and it centralizes the exit point)